pkg/agent/nodeportlocal/npl_agent_init.go

```diff
@@ -31,30 +34,110 @@ import (
+	podList, err := kubeClient.CoreV1().Pods(metav1.NamespaceAll).List(context.TODO(), metav1.ListOptions{
+		FieldSelector: "spec.nodeName=" + nodeName,
+	})
+	if err != nil {
+		return fmt.Errorf("error in fetching the Pods for Node %s: %s", nodeName, err.Error())
+	}
+
+	allNPLPorts := []rules.PodNodePort{}
+	for i := range podList.Items {
+		// For each Pod:
+		// check if a valid NPL Annotation exists for this Pod:
+		//   if yes, verifiy validity of the Node port, update the port table and add a rule to the
+		//   rules buffer.
+		annotations := podList.Items[i].GetAnnotations()
+		nplAnnotation, ok := annotations[nplk8s.NPLAnnotationKey]
+		if !ok {
+			continue
+		}
+		nplData := []nplk8s.NPLAnnotation{}
+		err := json.Unmarshal([]byte(nplAnnotation), &nplData)
+		if err != nil {
+			// if there's an error in this NPL Annotation, clean it up
+			err := nplk8s.CleanupNPLAnnotationForPod(kubeClient, &podList.Items[i])
+			if err != nil {
+				return err
+			}
+			continue
+		}
+
+		for _, npl := range nplData {
+			if npl.NodePort > portTable.EndPort || npl.NodePort < portTable.StartPort {
+				// invalid port, cleanup the NPL Annotation
+				if err := nplk8s.CleanupNPLAnnotationForPod(kubeClient, &podList.Items[i]); err != nil {
+					return err
+				}
+				break
+			} else {
+				allNPLPorts = append(allNPLPorts, rules.PodNodePort{
+					NodePort: npl.NodePort,
+					PodPort:  npl.PodPort,
+					PodIP:    podList.Items[i].Status.PodIP,
+				})
+			}
+		}
+	}
+
+	if len(allNPLPorts) > 0 {
+		if err := addRulesForNPLPorts(portTable, allNPLPorts); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
 // InitializeNPLAgent initializes the NodePortLocal (NPL) agent.
 // It initializes the port table cache to keep track of Node ports available for use by NPL,
 // sets up event handlers to handle Pod add, update and delete events.
 // When a Pod gets created, a free Node port is obtained from the port table cache and a DNAT rule is added to NAT traffic to the Pod's ip:port.
 func InitializeNPLAgent(kubeClient clientset.Interface, informerFactory informers.SharedInformerFactory, portRange, nodeName string) (*nplk8s.NPLController, error) {
 	start, end, err := util.ParsePortsRange(portRange)
 	if err != nil {
-		return nil, fmt.Errorf("something went wrong while fetching port range: %v", err)
+		return nil, fmt.Errorf("error while fetching port range: %v", err)
 	}
 	var ok bool
 	portTable, ok := portcache.NewPortTable(start, end)
 	if !ok {
-		return nil, errors.New("NPL port table could not be initialized")
+		return nil, errors.New("error in initializing NPL port table")
 	}
+
 	err = portTable.PodPortRules.Init()
 	if err != nil {
-		return nil, fmt.Errorf("NPL rules for pod ports could not be initialized, error: %v", err)
+		return nil, err
+	}
+
+	klog.Info("Will fetch Pods and generate NPL rules for these Pods")
+	if err := getPodsAndGenRules(kubeClient, portTable, nodeName); err != nil {
+		return nil, err
 	}
```

Comment:
I feel like this should be done in the `Run` method of the controller, after the caches are synced and before starting the workers.

Reply:
Yeah, agreed. Moved it. Thanks.